doc/tutorial.txt

```diff
@@ -256,6 +256,81 @@ inherit the affinity from the task level):
 	}
 }
 
+*** taskgroups
+
+* taskgroup: String. Can be specified at task level or phase level. Currently
+only SCHED_OTHER tasks are supported. An empty taskgroup value ("") is allowed
+and is interpreted as if there is no taskgroup name/value pair given.
+Pre-existing parts of taskgroups are preserved. In case that parts of
+taskgroups specified in the json file already existed, they are not deleted
+during the cleanup phase.
+A binary without taskgroup support ignores any taskgroup related name/value
+pairs.
+
+The following example sets up a task "thread1" to run in taskgroup "/tg1".
+
+"tasks" : {
+        "thread1" : {
+                "taskgroup" : "/tg1",
+                "run" : 10,
+                "sleep" : 10
+        }
+}
+
+The following example sets up a task "thread2" with three phases. "thread2"
+will run in taskgroup "/tg1/tg11" in phase1. Phase2 does not specify a
+taskgroup so "thread2" will continue to run in "/tg1/tg11". In phase3 "thread2"
+will run in taskgroup "/" (root).
+
+"tasks" : {
+        "thread2" : {
+                "phases" : {
+                        "phase1" : {
+                                "taskgroup" : "/tg1/tg11",
+                                "run" : 10,
+                                "sleep" : 10
+                        },
+                        "phase2" : {
+                                "run" : 10,
+                                "sleep" : 10
+                        },
+                        "phase3" : {
+                                "taskgroup" : "/",
+                                "run" : 10,
+                                "sleep" : 10
+                        }
+                }
+        }
+}
+
+Specifying taskgroups on task and phase level for a task should be avoided.
```

Comment:
That's actually easy to achieve with the two patches. I will merge them into the new version:

```commit 21d66e2f87de7fea1c6a32e0c418bcb1003c2d36
Author: Dietmar Eggemann <dietmar.eggemann@arm.com>
Date:   Fri Oct 19 12:11:19 2018 +0200

    Fix: Do not set taskgroup for thread explicitly
    
    In case a thread does not have phases the function parse_thread_data()
    creates a single phase for that thread. So the taskgroup configuration
    data for the thread is used for this single phase object which means we
    do not have to parse, create, set or remove the thread's taskgroup since
    it is done by this single phase object.
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app.c b/src/rt-app.c
index 9125f3601685..e0a92ac84d3d 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -843,7 +843,6 @@ void *thread_body(void *arg)
                        data->ind, policy_to_string(data->sched_data->policy),
                        data->sched_data->prio);
        set_thread_priority(data, data->sched_data);
-       set_taskgroup(data, data->taskgroup);
 
        /*
         * phase        - index of current phase in data->phases array
diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index 60270735f8c8..15379902989d 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -871,7 +871,6 @@ parse_thread_data(char *name, struct json_object *obj, int index,
        parse_cpuset_data(obj, &data->cpu_data);
        /* Scheduling policy */
        data->sched_data = parse_sched_data(obj, opts->policy);
-       data->taskgroup = parse_taskgroup_data(obj, data->sched_data, opts);
        /* initial delay */
        data->delay = get_int_value_from(obj, "delay", TRUE, 0);
 
diff --git a/src/rt-app_taskgroups.c b/src/rt-app_taskgroups.c
index b3f4a58e78b0..ebe092e8f01b 100644
--- a/src/rt-app_taskgroups.c
+++ b/src/rt-app_taskgroups.c
@@ -273,8 +273,6 @@ void create_taskgroups(rtapp_options_t *opts)
        for (i = 0; i < opts->nthreads; i++) {
                thread_data_t *tdata = &opts->threads_data[i];
 
-               create_taskgroup(tdata->taskgroup);
-
                for (y = 0; y < tdata->nphases; y++) {
                        phase_data_t *pdata = &tdata->phases[y];
 
@@ -312,9 +310,6 @@ void remove_taskgroups(rtapp_options_t *opts)
 
                        remove_taskgroup(pdata->taskgroup);
                }
-
-               remove_taskgroup(tdata->taskgroup);
-
        }
 }
 
diff --git a/src/rt-app_types.h b/src/rt-app_types.h
index e831b2f929a2..11204b0ba837 100644
--- a/src/rt-app_types.h
+++ b/src/rt-app_types.h
@@ -199,7 +199,6 @@ typedef struct _thread_data_t {
        FILE *log_handler;
 
        unsigned long delay;
-       taskgroup_t *taskgroup;
 } thread_data_t;
```

```Author: Dietmar Eggemann <dietmar.eggemann@arm.com>
Date:   Mon Oct 22 10:22:20 2018 +0200

    Fix: Check thread with phases for taskgroup data on thread level
    
    Specifying taskgroups on task and phase level for a task should be
    avoided. Otherwise, it is possible that a phase of a task runs in
    different taskgroups. For example, phase0 of task thread0 would run in
    "/tg1" in the first loop and in "/tg1/tg11" in the second loop.
    
    "tasks" : {
            "thread0" : {
                    "loop": 2,
                    "taskgroup" : "/tg1",
                    "phases" : {
                            "phase0" : {
                                    "run" : 10,
                                    "sleep" : 10
                            },
                            "phase1" : {
                                    "taskgroup" : "/tg1/tg11",
                                    "run" : 10,
                                    "sleep" : 10
                            }
                    }
            }
    }
    
    Taskgroup should only be specified on thread level if the thread does
    not contain one or more phases. Otherwise fail to start the run with
    an appropriate log message.
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index 15379902989d..0ba615e2e01c 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -795,12 +795,24 @@ static taskgroup_t *parse_taskgroup_data(struct json_object *obj, sched_data_t *
 
        return tg;
 }
+
+static int has_taskgroup_data(struct json_object *obj)
+{
+       char *name = get_string_value_from(obj, "taskgroup", TRUE, "");
+
+       return strlen(name) ? 1 : 0;
+}
 #else
 static taskgroup_t *parse_taskgroup_data(struct json_object *obj, sched_data_t *data,
                                         rtapp_options_t *opts)
 {
        return NULL;
 }
+
+static int has_taskgroup_data(struct json_object *obj)
+{
+       return 0;
+}
 #endif
 
 static void
@@ -883,6 +895,15 @@ parse_thread_data(char *name, struct json_object *obj, int index,
                assure_type_is(phases_obj, obj, "phases",
                                        json_type_object);
 
+               /*
+                * taskgroup should only be specified on thread level if the thread
+                * does not contain one or more phases.
+                */
+               if (has_taskgroup_data(obj)) {
+                       log_critical("Specify taskgroup in phase rather than thread");
+                       exit(EXIT_INV_CONFIG);
+               }
+
                log_info(PIN "Parsing phases section");
                data->nphases = 0;
                foreach(phases_obj, entry, key, val, idx) {
```